docs/adrs/adr007-data-validation.md

```diff
@@ -0,0 +1,145 @@
+custom events.
+
+[MSC2801][msc2801] provides reasons why events may not match the specification and the conclusion that
+clients must treat all event data as untrusted.
+
+Matrix events are described by the Matrix specification with [event schemas][eventSchemas]. This should be used as
+a reference to implement validation logic of the events used.
+
+Widgets may also load data from other data sources: localstorage, RTC channel, HTTP APIs. In general,
+we could not assume these events and data to be valid all the time and therefore must validate them.
+
+## Decision
+
+<!-- This section describes our response to these forces. It is stated in full sentences, with active voice. "We will ..." -->
+
+We validate all events and data that is received by widgets from different sources:
+
+For Matrix events (both from the Matrix specification and custom widget events):
+
+- `type` has expected event type, example: `m.room.name`, `net.nordeck.whiteboard`
+- `content` to be an `object` and to be defined
+- `content` is validated to match the schema
+
+For events and data from other data sources: we apply validation of data to match the schema.
+
+Schema based validation includes:
+
+- required fields are listed
+- optional and required fields values have the expected data types
+- field's primitive values match expected grammar: dates in ISO 8601, MXC URIs, whitelist values
+- field's object data is also validated according to schema
+
+We use several ways to validate incoming data.
+
+### Plain code
+
+We write validation logic in plain code when it is not complicated or we don't want to use a library, introducing another external dependency.
+
+[Example][examplePlainRelatesTo] that validates that event has [`m.relates_to` object][matrixEventsRelations] in the content:
+
+<!--prettier-ignore-->
+   ```ts
+    // ....
+
+    /**
+     * Validates that `event` has a valid structure for a
+     * {@link EventWithRelatesTo}.
+     * @param event - The event to validate.
+     * @returns True, if the event is valid.
+     */
+    export function isValidEventWithRelatesTo(
+      event: RoomEvent,
+    ): event is EventWithRelatesTo<string> {
+      if (!event.content || typeof event.content !== 'object') {
+        return false;
+      }
+    
+      const relatedEvent = event as EventWithRelatesTo<string>;
+    
+      if (
+        !relatedEvent.content['m.relates_to'] ||
+        typeof relatedEvent.content['m.relates_to'] !== 'object'
+      ) {
+        return false;
+      }
+    
+      if (
+        typeof relatedEvent.content['m.relates_to'].rel_type !== 'string' ||
+        typeof relatedEvent.content['m.relates_to'].event_id !== 'string'
+      ) {
+        return false;
+      }
+    
+      return true;
+    }
+   ```
+
+### Joi
+
+We prefer to use [Joi][joi] when we need to validate data with more complex structure. Joi provides powerful
+validation features.
```

Comment:
Maybe this should be more general "using a validation library" (compared to "plain code") and quote Joi as an example for validating JSON? I'm not sure what kinds of data Joi can validate, but another important example would be file types and structure of NeoBoard images and PDF uploads.

Reply:
Did some smaller updates to that. Consequences also has a link to `example-widget-mui` with examples how to use Joi.